src/rpc/RpcChannel.cpp

```diff
@@ -51,6 +52,70 @@ using namespace google::protobuf::io;
 namespace Hdfs {
 namespace Internal {
 
+
+class SaslOutputWrapper {
+public:
+    SaslOutputWrapper(SaslClient * client, RpcClient *rpcClient) {
+        saslClient = client;
+        this->rpcClient = rpcClient;
+    }
+
+    std::string wrap(std::string data) {
+        if (!saslClient->isPrivate() && !saslClient->isIntegrity())
+            return data;
+
+        std::string rawdata = saslClient->encode(data.c_str(), data.length());
+        msg.set_state(RpcSaslProto_SaslState_WRAP);
+        msg.set_token(rawdata);
+        int totalLen;
+        WriteBuffer buffer;
+        RpcRequestHeaderProto rpcHeader;
+        rpcHeader.set_callid(AuthProtocol::SASL);
+        rpcHeader.set_clientid(rpcClient->getClientId());
+        rpcHeader.set_retrycount(INVALID_RETRY_COUNT);
+        rpcHeader.set_rpckind(RPC_PROTOCOL_BUFFER);
+        rpcHeader.set_rpcop(RpcRequestHeaderProto_OperationProto_RPC_FINAL_PACKET);
+        RpcContentWrapper wrapper(&rpcHeader, &msg);
+        totalLen = wrapper.getLength();
+        buffer.writeBigEndian(totalLen);
+        wrapper.writeTo(buffer);
+
+        int length = buffer.getDataSize(0);
+        std::string outdata;
+        outdata.resize(length);
+        memcpy(&outdata[0], buffer.getBuffer(0), length);
```

Comment:
bad idea to use low-level functions for high-level C++ objects.:
If `buffer.getBuffer(0)` returns `char *`:
```
const char * buf = buffer.getBuffer(0)
std::string outdata(buf, buf + length)
```
or if not:
```
const char * buf = (const char *)buffer.getBuffer(0) // better use reinterpret_cast or static_cast.
std::string outdata(buf, buf + length)
```